### PR TITLE
[AUTO] AUTO inferences with CPU while the actual accelerator loads the network

### PR DESCRIPTION
**Pull request by mashoujiang**
### Details:
 - *this is a draft; AUTO inferences with CPU while the actual accelerator loads the network*

### Tickets:
 - *56054*

### CODE REVIEW

**Comment by mashoujiang** [comment]
openvino/inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:390 [ NOT_IMPLEMENTED ] Input image format FP16 is not supported yet... but FP16 is OK for GPU...

**Comment by mashoujiang** [comment]
This PR will change the behavior of AUTO, so lots of config CI tests should be fixed.

**Comment by mashoujiang** [comment]
Hi, @myshevts @ilya-lavrenov , I am trying to fix the CI issues for AUTO:CPU,GPU tests, but I couldn't reproduce in my servers which has i7-8700k with integrated GPU as the same with [ie-tests-linux-ubuntu18-gpu](https://openvino-ci.intel.com/job/private-ci/job/ie/job/ie-tests-linux-ubuntu18-gpu/43981/ ). I want to get the segment fault log but not sure how to get. do you have any suggestion about that CI failure? Thanks!

**Comment by myshevts** [comment]
to me the issue manifest itself in the scenarios when the test ends faster (and the Executable network is destroyed) before the thread completed the load.

as it manifests itself on the centOs as well - https://openvino-ci.intel.com/job/private-ci/job/ie/job/ie-tests-linux-centos76-gpu/27270/consoleFull#1680878394603a3860efd2da540410cf02, I think this is fair issue.

Likely the reason is how the google tests are operating (e.g. maybe they are <not> destroying smth unlike real apps).
Anyway likely we need to keep a shred_ptr to the GPU plugin in the thread that does the GPU network load... I think the reason of segfualt is that the thread touches smth that is already destroyed

**Comment by myshevts** [comment]
also let's kindly ask @apankratovantonp  to look at the code as well, just to make sure the issue is not related to how you use the taskeExecutor - https://github.com/openvinotoolkit/openvino/pull/5944/files#diff-a42d86638395609ea0f2965129ad627d6554395bf362e9e8cb1e88ce1ab753a9R96

BTW, was the code also failing in the intial ver (without taskeExeutor and with pure std::async)?

**Comment by mashoujiang** [comment]
> also let's kindly ask @apankratovantonp to look at the code as well, just to make sure the issue is not related to how you use the taskeExecutor - https://github.com/openvinotoolkit/openvino/pull/5944/files#diff-a42d86638395609ea0f2965129ad627d6554395bf362e9e8cb1e88ce1ab753a9R96
> 
> BTW, was the code also failing in the intial ver (without taskeExeutor and with pure std::async)?

Thanks Maxim, the initial ver is using std::thread, which also has this segment fault. I didn't try std::async(), maybe that is good enough, will have a try. 
And I tried sync mode by [waiting actual network](https://github.com/openvinotoolkit/openvino/pull/5944/files#diff-750d1c33e67b07ac0387961ac068deeeaa0e041e6d9b6cbdcf9f17bb96151201R34) which also has segment fault.

I see comments from @apankratovantonp , I will fix and try more. also I give you some comments in #5839 which I am trying to update.
Thanks!

File: inference-engine/src/auto_plugin/auto_plugin.cpp
Comment (mashoujiang) [reply inference-engine/src/auto_plugin/auto_plugin.cpp:110]: Yes, thanks, will update.

File: inference-engine/src/auto_plugin/auto_exec_network.cpp
Comment (mashoujiang) [reply inference-engine/src/auto_plugin/auto_exec_network.cpp:24]: I think the Executor lifetime is bound to AutoExecutableNetwork by [_cpuExecutor](https://github.com/openvinotoolkit/openvino/pull/5944/files#diff-750d1c33e67b07ac0387961ac068deeeaa0e041e6d9b6cbdcf9f17bb96151201R21)

File: inference-engine/src/auto_plugin/auto_plugin.cpp
Comment (mashoujiang) [reply inference-engine/src/auto_plugin/auto_plugin.cpp:96]: Thanks for suggestions! I will check again~

File: inference-engine/src/auto_plugin/auto_exec_network.cpp
Comment (apankratovantonp) [reply inference-engine/src/auto_plugin/auto_exec_network.cpp]: `cpuTask` task is run in executor that was created in plugin::LoadNetworkImpl. As it is gotten from the executor manager the executor and its threads are bound to executor manager lifetime not to executor network.

Comment (mashoujiang) [reply inference-engine/src/auto_plugin/auto_exec_network.cpp]: I understand what you mean, `cpuTask ` shared ptr will run in `IStreamsExecutor` (that created by executor manager), but the shared cpuTask will not expire because it is refered in AutoExecutableNetwork, we could get_future().get() in AutoExecutableNetwork, is that right?
Sorry for the update code since rebasing to master branch (fix conflict) and trying std::async(), so code is changed, sorry for that.

File: inference-engine/src/auto_plugin/auto_exec_network.cpp
Comment (myshevts) [inline inference-engine/src/auto_plugin/auto_exec_network.cpp:23]: seemingly the future is always valid.
waht @apankratovantonp was trying to project that the .get() below may rethrow exception if the network loading to the CPU fails. Notice that unlike actual accelerator load, we may catch this exception (as CPU is a "helper" device only).

Comment (mashoujiang) [reply inference-engine/src/auto_plugin/auto_exec_network.cpp:23]: Thanks Maxim, well, I have used std::async() instead of Executor with tasks, it seems no failure now, all CI has passed. Also, I fixed some logic error, I believe the segment fault is solved (maybe it caused by the wrong used task/executor/future).
This PR is still missing some implementation you mentioned in email and tickets, I will continue to update it, and it is not completed yet. Of course, I will be happy if you give me any suggestion during my development, :) .  Thanks!

Comment (myshevts) [reply inference-engine/src/auto_plugin/auto_exec_network.cpp:23]: seems like your std::async impl is doing something very differently, as I don't see the debug printfs in the succesful tests log...
notice that the test may also pass becuase the stad::async may run serially (as you didn't force to run it in parallel).
Finally, as the original std:therad based impl also filaed, likely the reason is not the async method itslf (whther it is therad/packed_task,executor/etc) but inaccurate capture  of e.g. 'this' (as opposite to shared_ptr) etc.

I strongly sughest to remove the capturing by value in the lambdas and carefully review what is captured and whta is lifetime of this (and if it is needed to be aligned)

Comment (mashoujiang) [reply inference-engine/src/auto_plugin/auto_exec_network.cpp]: Yes, the std::async logic is different from init task impl because the original hot_swap logic is wrong.
I have force the std::async to run in parallel now.
I have changed capturing by reference, do you mean that? (you:"remove the capturing by value in the lambdas")
I think the code is ready for review now, :)

**Comment by mashoujiang** [comment]
The segment fault or core dump may caused by ~ICore() destructor earlier than Auto plugin, should we make sure plugin destructed later than ICore?
`~ICore()
!!! DEBUG: GPU was loaded !!!
Debug: ~AutoInferencePlugin()`

File: inference-engine/src/auto_plugin/auto_plugin.cpp
Comment (mashoujiang) [inline inference-engine/src/auto_plugin/auto_plugin.cpp:83]: this actually does not solve segment fault, please ignore this change.

Comment (myshevts) [reply inference-engine/src/auto_plugin/auto_plugin.cpp:83]: I was referrring to the lambda capturing being too broad (i.e [=])  not function args/
capturing everything by value (incl 'this') is dangerous, again you have to capture the right things only and only as shared_ptr, e.g. the GetCore should be probably cllaed outside lambda, wrapped to the shared_ptr, and the ptr can be captured by value then, this is exactly you are asking from @ilya-lavrenov to advice in another comment

Comment (mashoujiang) [reply inference-engine/src/auto_plugin/auto_plugin.cpp:83]: GetCore() is already a raw pointer, that is `InferenceEngine::ICore* core = GetCore();`, I am not sure if it is a good idea to wrap this to shared_ptr. But I agree that we should capture explicit, I will change that. Thanks~

Comment (mashoujiang) [reply inference-engine/src/auto_plugin/auto_plugin.cpp:83]: and the question I am asking ilya is that if it is OK that ~ICore() happened before any plugin. In AUTO case, ~AUTOPLUGIN() definitely will still call ICore() object. do we have any way to make the lifetime of ICore() as the same with plugin?

**Comment by myshevts** [comment]
> may need @ilya-lavrenov help to analysis since you may be more familiar with this IE Core framework.

as already pointed to prevent ICore being destroyed, you should capture that outside the lambdas and store the shared_ptr...
Not sure if the current plugin interface is enough though, as the GetCOre return the plain pointer, not shared,

**Comment by mashoujiang** [comment]
> > may need @ilya-lavrenov help to analysis since you may be more familiar with this IE Core framework.
> 
> per @ilya-lavrenov, the Core holds the plugins (and not vice versa, to avoid cyclic references), so the Core should not die, before all the plugins it holds have been desroyed. Again, maybe this is specifics of the Google test? E.g. if the it is using some PluginCache...

Thanks Maxim, you're right, the Core is alive before plugin destroied in GTest. (BTW, Core is destoried eariler than plugin in beanchmark_app). 
The AUTO tests in CI are not stable, you can see above CI tests log, I got **52 pass green** in one time. But without any code change, the newest CI failed again (no segment fault), and I could not reproduce in my machine....

File: inference-engine/src/auto_plugin/auto_plugin.cpp
Comment (myshevts) [inline inference-engine/src/auto_plugin/auto_plugin.cpp:113]: the comment is incorrect :)

**Review by myshevts — COMMENTED** [review]
Before merging, make sure to remove the debug output

File: inference-engine/src/auto_plugin/auto_plugin.cpp
Comment (mashoujiang) [reply inference-engine/src/auto_plugin/auto_plugin.cpp:113]: yes, will modify. BTW do you think should we use the default "std::launch::async | std::launch::deferred" ?

File: inference-engine/src/auto_plugin/auto_exec_network.cpp
Comment (myshevts) [inline inference-engine/src/auto_plugin/auto_exec_network.cpp:24]: if CPU loading failed, and the CPU is NOT the accelerator, we should catch the exception (whihc the future::get() will re-throw).
this is captured as " // TODO: FIXME: revert the exception handling logic back to gracefully handle LoadNetwork failures" in another place but should be handled here I guess

Comment (mashoujiang) [reply inference-engine/src/auto_plugin/auto_exec_network.cpp]: agreed, addressed, thanks!
BTW, I want to keep debug log until no any other issues since may still need to debug if code change.

**Comment by mashoujiang** [comment]
Hi @myshevts @ilya-lavrenov , would you or someone else please help to review this PR?

**Comment by myshevts** [comment]
I made a review and has conditionally approved that (pending the preso to the Arch Forum on ratifying the behavior)

**Comment by mashoujiang** [comment]
@myshevts some CI tests could not pass, like hddl-plugin and kmb-plugin, I believe those plugins will use the GetCore() and report the error like below, and they belong to different repo: [hddl-plugin](https://gitlab-icv.inn.intel.com/inference-engine/hddl-plugin), [kmb-plugin](https://gitlab-icv.inn.intel.com/inference-engine/kmb-plugin)
 **/home/jenkins/agent/workspace/private-ci/ie/build-linux-ubuntu20/b/repos/hddl-plugin/src/hddl_executable_network.cpp:77:9: error: could not convert ���((vpu::HDDLPlugin::ExecutableNetwork*)this)->vpu::HDDLPlugin::ExecutableNetwork::_core��� from ���const InferenceEngine::ICore*��� to ���std::shared_ptr<InferenceEngine::ICore>���**

**Comment by myshevts** [comment]
> @myshevts some CI tests could not pass, like hddl-plugin and kmb-plugin, I believe those plugins will use the GetCore() and report the error like below, and they belong to different repo: [hddl-plugin](https://gitlab-icv.inn.intel.com/inference-engine/hddl-plugin), [kmb-plugin](https://gitlab-icv.inn.intel.com/inference-engine/kmb-plugin)
> 

yes, you have to change the impl in these plugins as well

**Comment by ilya-lavrenov** [comment]
We need to update plugins and not add new methods

**Review by myshevts — APPROVED** [review]
pls merge togther with chnages to the HDDL-plugin and write down a broadcast e-meil asking people to rebase their PRs to avoid CI failure

**Pull request merged**
